ports/esp8266/esp_mphal.c

```diff
@@ -87,23 +82,39 @@ void mp_hal_debug_str(const char *str) {
 #endif
 
 void mp_hal_stdout_tx_str(const char *str) {
+    const char *last = str;
     while (*str) {
-        mp_hal_stdout_tx_char(*str++);
+        uart_tx_one_char(UART0, *str++);
     }
+    mp_uos_dupterm_tx_strn(last, str - last);
 }
 
 void mp_hal_stdout_tx_strn(const char *str, uint32_t len) {
+    const char *last = str;
     while (len--) {
-        mp_hal_stdout_tx_char(*str++);
+        uart_tx_one_char(UART0, *str++);
     }
+    mp_uos_dupterm_tx_strn(last, str - last);
```

Comment:
Well, so we have `len` from the start, no need to calculate it.

Reply:
But len is zero at this point due to the loop.  And I felt it important to output to the UART before the dupterm, in case the latter had errors at least the data got to the UART first.

Reply:
D'oh, right, looks good then.